### ISSUE TITLE
feat(gpu): Full GPU inference & training — end-to-end VRAM-resident forward/backward pass

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/ATLAS#3]
## Overview

ATLAS currently has **GPU-accelerated CPU inference**: the 7 weight projection matmuls per layer run via `atlas_matmul_f32` CUDA kernels and weight matrices are pre-pinned in VRAM via `GpuMatrix`. This gives ~30 tok/s on a T4 (5.8× over CPU).

But it is **not** end-to-end GPU inference. The hidden state `x` lives in CPU RAM and crosses the PCIe bus 211 times per token (once per matmul call). RMSNorm, RoPE, softmax, SwiGLU, attention scores, residual adds, and gradient accumulation all run on CPU.

This issue tracks the work to make ATLAS fully GPU-resident: hidden state stays in VRAM for the entire forward (and backward) pass, with only the token embedding uploaded and the logits downloaded per step.

---

## Current state

| Component | Device | Notes |
|-----------|--------|-------|
| Weight matrices | **VRAM** ✅ | `GpuMatrix` pre-uploads at load |
| Matrix multiplications | **GPU** ✅ | `atlas_matmul_f32` sm_75 CUDA kernel |
| Hidden state `x` | **CPU** ❌ | `Vec<f32>`, bounces PCIe 211×/token |
| RMSNorm | **CPU** ❌ | Pure Rust loop |
| RoPE | **CPU** ❌ | Per-head in-place |
| Attention scores + softmax | **CPU** ❌ | Dot products + stable softmax |
| SwiGLU element-wise | **CPU** ❌ | `silu(gate) * up` fusion |
| Residual adds | **CPU** ❌ | `+=` loop |
| KV cache | **CPU** ❌ | `Vec<f32>` |
| GradTape / AdamW | **CPU** ❌ | No GPU path wired up |

**Throughput gap:** ~30 tok/s now vs ~150–300 tok/s achievable on T4 with full GPU inference.

---

### Phase 1 — New CUDA kernels (`kernels/ops.cu`)

Add the missing element-wise and reduction kernels. All follow the same zero-dependency pattern as `matmul.cu`.

```
kernels/ops.cu
```

| Kernel | Signature | Notes |
|--------|-----------|-------|
| `atlas_rmsnorm_f32` | `(x, w, out, n, eps)` | Parallel reduction over `n` elements |
| `atlas_rope_f32` | `(x, pos, head_dim, theta, n_heads)` | In-place RoPE rotation |
| `atlas_silu_mul_f32` | `(gate, up, out, n)` | Fused SwiGLU: `silu(gate[i]) * up[i]` |
| `atlas_vec_add_f32` | already declared in FFI — wire up | Residual add |
| `atlas_attn_softmax_f32` | `(scores, seq_len, n_heads)` | Per-head softmax, numerically stable |
| `atlas_attn_scores_f32` | `(q, k_cache, scores, head_dim, seq_len)` | Batched dot products for attention |
| `atlas_attn_output_f32` | `(scores, v_cache, out, head_dim, seq_len)` | Weighted value accumulation |

Expose each via `extern "C"` in `atlas-tensor/src/lib.rs` FFI block.

---

### Phase 2 — `GpuVec`: GPU-resident activation buffer (`atlas-tensor`)

Add a lightweight GPU buffer for activations (distinct from `GpuMatrix` which is for static weights):

```rust
/// Mutable GPU buffer for transient activations (hidden states, KV cache, etc.)
pub struct GpuVec {
    buf: gpu::GpuBuf,
    pub len: usize,
}

impl GpuVec {
    pub fn zeros(len: usize) -> Option<Self>;
    pub fn from_slice(data: &[f32]) -> Option<Self>;
    pub fn download(&self) -> Vec<f32>;
    // In-place ops dispatching to GPU kernels:
    pub fn rmsnorm_inplace(&mut self, w: &GpuVec, eps: f32);
    pub fn add_inplace(&mut self, other: &GpuVec);
    pub fn scale_inplace(&mut self, s: f32);
}
```

Falls back to CPU allocation if CUDA not available — zero API breakage.

---

### Phase 3 — GPU-resident forward pass (`atlas-model`)

Refactor `TransformerBlock::forward_token` and `OlmoModel::forward_one` to pass `GpuVec` instead of `Vec<f32>`:

```
forward_one(token: u32) → Vec<f32>
  │
  ├─ embed_token(token) → upload to GpuVec  [1× H2D, ~2 KB]
  │
  ├─ for each layer:
  │   ├─ rmsnorm_inplace (GPU kernel)
  │   ├─ Wq·x, Wk·x, Wv·x  (GpuMatrix::sgemm, stays in VRAM)
  │   ├─ rope_inplace (GPU kernel)
  │   ├─ KV cache write (VRAM → VRAM copy)
  │   ├─ attn_scores + softmax (GPU kernels)
  │   ├─ attn_output (GPU kernel)
  │   ├─ Wo·attn (GpuMatrix::sgemm)
  │   ├─ residual add (GPU kernel)
  │   ├─ rmsnorm_inplace (GPU kernel)
  │   ├─ W_gate·x, W_up·x (GpuMatrix::sgemm)
  │   ├─ silu_mul_inplace (GPU kernel)
  │   ├─ W_down·h (GpuMatrix::sgemm)
  │   └─ residual add (GPU kernel)
  │
  ├─ final rmsnorm (GPU kernel)
  ├─ lm_head (GpuMatrix::sgemm)
  └─ download logits [1× D2H, ~192 KB]
```

**PCIe transfers per token: 2** (down from 211 today).

CPU fallback: if `GpuVec::from_slice()` returns `None` (no CUDA), `forward_one` falls back to the existing `Vec<f32>` path automatically.

---

### Phase 4 — GPU KV cache

Replace `KvCache { keys: Vec<f32>, values: Vec<f32> }` with:

```rust
struct KvCache {
    keys:   Option<GpuVec>,   // VRAM resident
    values: Option<GpuVec>,   // VRAM resident
    // CPU fallback:
    keys_cpu:   Vec<f32>,
    values_cpu: Vec<f32>,
    ...
}
```

KV cache reads/writes become VRAM→VRAM copies (no PCIe). Critical for long-context performance.

---

### Phase 5 — GPU training (`atlas-grad`, `atlas-optim`)

Wire up the existing FFI declarations (`atlas_vec_add_f32`, `atlas_scale_f32`) into the training stack:

1. `GradTape::push()` — call `.to_cuda()` on tensors if `cuda_available()`
2. `GradTape::matmul()` — already dispatches to `Tensor::matmul()` which has GPU path; just needs CUDA tensors in
3. `GradTape::backward()` — gradient tensors move to GPU via `.to_cuda()`
4. `AdamW::step()` — parameter and moment updates use `atlas_vec_add_f32` + `atlas_scale_f32` kernels
5. `SftTrainer` — call `model.to_cuda()` after construction (new method that calls `.to_cuda()` on all `GpuMatrix` weight buffers — already done — and `GradTape` tensors)

---

## Expected outcomes

| Metric | Current | After Phase 1–4 | After Phase 5 |
|--------|---------|-----------------|---------------|
| Inference tok/s (T4) | ~30 | **~150–300** | — |
| PCIe transfers/token | 211 | **2** | — |
| GPU util during inference | 71–80% | **>95%** | — |
| Training steps/s | CPU-only | CPU-only | **GPU** |
| GPU util during training | 0% | 0% | **>90%** |

---

## Implementation order

1. `kernels/ops.cu` — all new kernels in one file
2. `atlas-tensor`: FFI declarations + `GpuVec` struct
3. `atlas-model`: GPU-resident `forward_one` + GPU KV cache
4. `atlas-grad` / `atlas-optim`: wire `.to_cuda()` into training loop
5. Update `atlas-cli infer` to report PCIe transfer count and actual GPU util
6. Update `atlas bench` with GPU inference throughput benchmark
7. All 400 existing tests must continue to pass (CPU fallback path)

---

## Hardware target

Primary: NVIDIA Tesla T4 (sm_75, 15 GB VRAM, 300 GB/s bandwidth) — `astra-01`
The `find_nvcc()` and `gpu_arch()` functions in `atlas-tensor/build.rs` already handle auto-detection.

---

*Opened by FETCH-AGI. Current state: `GpuMatrix` (weight pre-upload + matmul dispatch) merged at `7d7af09`. This issue covers the remaining work to make every op GPU-resident.*

**Comment by web3guru888** [comment]
Closed by v3.0.0: GPU ops kernels (rmsnorm/rope/silu_mul) added (kernels/matmul.cu), GpuVec activation buffer (atlas-tensor), sgemm_vec zero-copy path, GPU AdamW via atlas_adamw_step, GPU-resident forward pass (forward_one_gpu: 2 PCIe transfers/token vs 211). Full GPU training via step_gpu(). Benchmark: see build-report-v3-final.md. Also includes rmsnorm inter-warp reduction bugfix (smem[0] broadcast instead of __shfl_sync).
